File: dynamic-programming-hard/312.戳气球.py
#
# @lc app=leetcode.cn id=312 lang=python3
#
# [312] 戳气球
#
from typing import List


# @lc code=start
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        n = len(nums)
        val = [1] + nums + [1]
        dp = [[0] * (n + 2) for _ in range(n + 2)]
        for i in range(n, -1, -1):
            for j in range(i+1, n+2):
                for k in range(i+1, j):
                    dp[i][j] = max(dp[i][j], dp[i][k]+dp[k]
                                   [j]+val[i]*val[k]*val[j])
        return dp[0][-1]
# @lc code=end

File: divide-and-conquer-hard/312.戳气球.cpp
/*
 * @lc app=leetcode.cn id=312 lang=cpp
 *
 * [312] 戳气球
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
#define it insert
#define pob pop_back
#define pub push_back
#define all(v) (v).begin(), (v).end()
using LL = long long;
using VI = vector<int>;
using VVI = vector<vector<int>>;
using PII = pair<int, int>;
using PIL = pair<int, LL>;
using PLL = pair<LL, LL>;
const int MOD = 1e9 + 7;
const int INF = 1e9 + 7;
class Solution {
public:
    VVI memo;
    VI v;
    int maxCoins(vector<int> &nums) {
        int n = nums.size();
        v = VI(nums.begin(), nums.end());
        v.it(v.begin(), 1);
        v.pub(1);
        memo.resize(n + 2, VI(n + 2, -1));
        return solve(0, n + 1);
    }
    int solve(int l, int r) {
        if (l >= r - 1) {
            return 0;
        }
    }
};
// @lc code=end
